src/auth/AuthEpics.js

```diff
@@ -18,107 +18,213 @@ import {
 
 export const login = (action$, store, call = (fn, ...args) => fn(...args)) => {
   return action$.ofType(LOGIN)
-    .mergeMap(({username, password, token, tenant}) => {
-      const state = store.getState();
-      const {authUrl} = state.configReducer;
-      const headers = {
-        'Content-Type': 'application/json'
-      };
-      const data = {
-        auth: {}
-      };
-
-      if (username !== undefined && password !== undefined) {
-        data.auth.passwordCredentials = {
-          username,
-          password
+    .mergeMap(({username, password, token, tenant, tenantId}) => {
+        const state = store.getState();
+        const {keystoneVersion, authUrl} = state.configReducer;
+        const headers = {
+          'Content-Type': 'application/json'
         };
-      } else if (token) {
-        data.auth.token = {
-          id: token,
+        const data = {
+          auth: {}
         };
-      }
 
-      if (tenant) {
-        data.auth.tenantName = tenant;
-      }
-      return call(
-        post,
-        `${authUrl}/tokens`,
-        headers,
-        data,
-      )
-        .flatMap(response => Observable.concat(
-          Observable.of(loginSuccess(
-            response.response.access.token.id,
-            response.response.access.token.expires,
-            response.response.access.token.tenant,
-            response.response.access.user,
-          )),
-          Observable.of({type: FETCH_TENANTS})),
-        )
-        .catch(error => Observable.of(loginFailure(parseXHRError(error))));
+        if (keystoneVersion === 'v2') {
+          if (username !== undefined && password !== undefined) {
+            data.auth.passwordCredentials = {
+              username,
+              password
+            };
+          } else if (token) {
+            data.auth.token = {
+              id: token
+            };
+          }
+
+          if (tenant) {
+            data.auth.tenantName = tenant;
+          }
+
+          return call(
+            post,
+            `${authUrl}/tokens`,
+            headers,
+            data
+          )
+            .flatMap(response => Observable.concat(
+              Observable.of(loginSuccess(
+                response.response.access.token.id,
+                response.response.access.token.expires,
+                response.response.access.token.tenant,
+                response.response.access.user
+              )),
+              Observable.of({type: FETCH_TENANTS}))
+            )
+            .catch(error => Observable.of(loginFailure(parseXHRError(error))));
+
+        } else if (keystoneVersion === 'v3') {
```

Comment:
Please add else for this if.
We need to show some error when `keystoneVersion` is undefined or have wrong value

Reply:
done